zeth-contracts/contracts/BaseMixer.sol

```diff
@@ -111,81 +111,42 @@ contract BaseMixer is MerkleTreeMiMC7, ERC223ReceivingContract {
     // This function processes the primary inputs to append and check the root and nullifiers in the primary inputs (instance)
     // and modifies the state of the mixer contract accordingly
     // (ie: Appends the commitments to the tree, appends the nullifiers to the list and so on)
-    function assemble_root_and_nullifiers_and_append_to_state(uint[] memory primary_inputs) internal {
-        // 1. We re-assemble the full root digest from the 2 field elements it was packed into
-        uint256[] memory digest_inputs = new uint[](2);
-        digest_inputs[0] = primary_inputs[0];
-
+    function check_mkroot_nullifiers_hsig_append_nullifiers_state(
+        uint[2][2] memory vk,
+        uint[] memory primary_inputs) internal {
+        // 1. We re-assemble the full root digest and check it is in the tree
         require(
-            roots[bytes32(digest_inputs[0])],
+            roots[bytes32(primary_inputs[0])],
             "Invalid root: This root doesn't exist"
         );
 
-        // 2. We re-assemble the nullifiers (JSInputs)
+        // 2. We re-assemble the nullifiers (JSInputs) and check they were not already seen
+        bytes32[jsIn] memory nfs;
+        uint256[] memory digest_inputs = new uint[](2);
         for(uint i = 1; i < 1 + 2*jsIn; i += 2) {
             digest_inputs[0] = primary_inputs[i];
             digest_inputs[1] = primary_inputs[i+1];
-            bytes32 current_nullifier = Bytes.sha256_digest_from_field_elements(digest_inputs);
+            nfs[(i-1)/2] = Bytes.sha256_digest_from_field_elements(digest_inputs);
```

Comment:
`(i-1)/2` why use this as index?

Reply:
This was to avoid using two different counters as we need one for digest_inputs and one for the nullifiers (or other variables).

We have 2 digest_inputs per nullifier (or other variable) so we need to divide $i$ by 2 to obtain the correct range for the nullifier index.
As the root is at index 0, i is odd ( 1, 3, 5... ) so we also need to remove 1 (not to rely on floor/ceil of the "/" operation).

Reply:
It should be more efficient to have a second counter actually (as we would have a C_variable_declaration + jsin*C_addition instead of jsin*(C_addition + C_division)  and C_var_declaration < C_division ).
 something like:

`
uint nf_count;\\
for(uint i = 1; i < 1 + 2*jsIn; i += 2) {\\
            digest_inputs[0] = primary_inputs[i];\\
            digest_inputs[1] = primary_inputs[i+1];\\
            bytes32 current_nullifier = Bytes.sha256_digest_from_field_elements(digest_inputs);\\
            nfs[nf_count] = Bytes.sha256_digest_from_field_elements(digest_inputs);\\
            nf_count = nf_count + 1;`

Reply:
Can you please open a ticket for this one @rrtoledo, it might also be relevant to @dtebbs for some of the work carried out on the contracts.

Reply:
Already did !

Reply:
I assume it is #133 ? Please reference the related PR and issues together so that we can easily track the discussions down